lib/jsonrpc/websocket.go

```diff
@@ -477,9 +511,22 @@ func (c *wsConn) handleWsConn(ctx context.Context) {
 			c.handleFrame(ctx, frame)
 			go c.nextMessage()
 		case req := <-c.requests:
+			c.writeLk.Lock()
 			if req.req.ID != nil {
+				if c.incomingErr != nil { // No conn?, immediate fail
```

Comment:
Interestingly, whenever gorilla detects the connection is closed, it closes `c.incoming`. Didn't expect that but found it while making some tests. So we never assume after reading `<-c.requests` that we have a good conn. `c.incomingErr` is set again to `nil` when conn is reestablished (line 490)